README.md
# Type Assertion And Predicate (`Taap`)

A lightweight zero dependency Typescript type checking library that coerces unknown value types. More concise and accurate than `typeof` checks or similar type checking methods.

<h3 align="center">
  <a href="https://schuchard.github.io/taap/" target="_blank">📚 Docs<a>
  &nbsp;
  <a href="https://stackblitz.com/edit/taap" target="_blank">⚡ Demo</a>
  &nbsp;
  <a href="https://github.com/schuchard/taap" target="_blank">💻 Repo</a>
</h3>

<p align="center">
  <a href="https://schuchard.github.io/taap/" target="_blank">
    <img alt="github-workflows" src="https://github.com/schuchard/taap/workflows/Release/badge.svg">
  </a>
  <a href="#badge" target="_blank">
    <img alt="semantic-release" src="https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg">
  </a>
  <a href="https://github.com/prettier/prettier" target="_blank">
    <img alt="styled with prettier" src="https://img.shields.io/badge/styled_with-prettier-ff69b4.svg">
  </a>
</p>

<p align="center">
  <a href="https://bundlephobia.com/result?p=taap" target="_blank">
    <img alt="bundlephobia" src="https://img.shields.io/bundlephobia/minzip/taap">
  </a>
  <a href="https://www.npmjs.com/package/taap" target="_blank">
    <img alt="npm latest version" src="https://img.shields.io/npm/v/taap/latest.svg">
  </a>
  <a href="https://www.npmjs.com/package/taap" target="_blank">
    <img alt="npm next version" src="https://img.shields.io/npm/v/taap/beta.svg">
  </a>
</p>

## 🛠 Getting Started

#### Install dependency

```bash
npm i taap
```

#### Import and use methods

```ts
import { isArray } from 'taap';

const maybeArray: unknown = [];

if (typeof maybeArray === 'array') {
  🚫 typeof [] === 'object'
} else if (isArray(maybeArray)) {
  // `maybeArray` is now of type: `any[]`
  ✅ maybeArray.push(1);
}
```

#### Optionally supply a return type

```ts
import { isArray } from 'taap';

const maybeArray: unknown = [1, 2, 3];

if (isArray<number>(maybeArray)) {
  // `maybeArray` is now of type: `number[]`
  maybeArray.filter((x) => x > 1);
}
```

## 🔭 Available Methods

### <a href="https://schuchard.github.io/taap/" target="_blank">📚Full documentation<a>

Supports optional generic return type:

- [`isArray()`](https://schuchard.github.io/taap/globals.html#isarray)
- [`isObject()`](https://schuchard.github.io/taap/globals.html#isobject)
- [`isFunction()`](https://schuchard.github.io/taap/globals.html#isfunction)

Fixed return type:

- [`isString()`](https://schuchard.github.io/taap/globals.html#isstring)
- [`isDate()`](https://schuchard.github.io/taap/globals.html#isdate)
- [`isRegExp()`](https://schuchard.github.io/taap/globals.html#isregexp)
- [`isBoolean()`](https://schuchard.github.io/taap/globals.html#isboolean)
- [`isNumber()`](https://schuchard.github.io/taap/globals.html#isnumber)
- [`isBigInt()`](https://schuchard.github.io/taap/globals.html#isbigint)
- [`isNull()`](https://schuchard.github.io/taap/globals.html#isnull)
- [`isUndefined()`](https://schuchard.github.io/taap/globals.html#isundefined)

Other:

- [`getType()`](https://schuchard.github.io/taap/globals.html#gettype)


CHANGELOG.md
# [1.2.0](https://github.com/schuchard/taap/compare/v1.1.0...v1.2.0) (2020-07-10)


### Features

* **docs:** update with link to repo ([1f527a2](https://github.com/schuchard/taap/commit/1f527a25ab5241b8c6bff633f79ce6ad9612d664))

# [1.1.0](https://github.com/schuchard/taap/compare/v1.0.3...v1.1.0) (2020-07-07)


### Features

* add BigInt support ([a8f934c](https://github.com/schuchard/taap/commit/a8f934c9e1b54cc2eec2c62a77c930e0431d8522))

## [1.0.3](https://github.com/schuchard/taap/compare/v1.0.2...v1.0.3) (2020-07-07)


### Bug Fixes

* add homepage ([ca146ac](https://github.com/schuchard/taap/commit/ca146ac2cb89ca8cf86748bb5f08187a1c085747))

## [1.0.2](https://github.com/schuchard/taap/compare/v1.0.1...v1.0.2) (2020-07-07)


### Bug Fixes

* **ci:** docs deploy ([2022caa](https://github.com/schuchard/taap/commit/2022caa7ca9dfe731f9658cb9eeec3f714caf606))

## [1.0.1](https://github.com/schuchard/taap/compare/v1.0.0...v1.0.1) (2020-07-07)


### Bug Fixes

* **ci:** doc deploy ([758b817](https://github.com/schuchard/taap/commit/758b8170580c358047045ce286dd76c89b39ca09))
* **ci:** doc deploy ([1bf1983](https://github.com/schuchard/taap/commit/1bf19831ad741a44fe4bcffa973c55d99e0ebff0))
* **ci:** docs deploy ([4897c9a](https://github.com/schuchard/taap/commit/4897c9a58bbcc178d39c1bdfa0d46d94689a1f94))
* **ci:** docs deploy ([0815857](https://github.com/schuchard/taap/commit/08158573c064acf362bf18ef5e0c56420ea4424b))
* **ci:** docs deploy ([38bb42e](https://github.com/schuchard/taap/commit/38bb42e1bf0cc33016f07421e29563c08689557f))
* **ci:** docs deploy ([fe9e070](https://github.com/schuchard/taap/commit/fe9e070d0d82082e5d03014b89f266f56c2d0010))
* **ci:** docs deploy ([0adc318](https://github.com/schuchard/taap/commit/0adc318383b7667b90b00ff6d369969517b1341d))
* **ci:** docs deploy ([e154590](https://github.com/schuchard/taap/commit/e1545906355463bf41336e1619e4d395f836fe89))
* **ci :** docs deploy ([0a0fe3f](https://github.com/schuchard/taap/commit/0a0fe3f42217ad78bf1f8bdac56642a4e265c1ab))

# 1.0.0 (2020-07-07)


### Features

* api, tests, ci/cd scripts, docs ([0badc6d](https://github.com/schuchard/taap/commit/0badc6df863314e3ace04ce4b1ce848a550d52b4))
* init and update project, add src files and specs ([6d6cc15](https://github.com/schuchard/taap/commit/6d6cc15767bfbfef82c2bda1de0d78843fc76650))

# [1.0.0-beta.10](https://github.com/schuchard/taap/compare/v1.0.0-beta.9...v1.0.0-beta.10) (2020-07-07)


### Features

* **docs:** deploy docs site ([83f4ccc](https://github.com/schuchard/taap/commit/83f4ccc37201d5e70b8ea859373604a6c8c827ba))

# [1.0.0-beta.9](https://github.com/schuchard/taap/compare/v1.0.0-beta.8...v1.0.0-beta.9) (2020-07-07)


### Features

* add TaapType type for getting type info ([37d2be5](https://github.com/schuchard/taap/commit/37d2be5aa2cf3daa497d58303265bd2cffef21f6))

# [1.0.0-beta.8](https://github.com/schuchard/taap/compare/v1.0.0-beta.7...v1.0.0-beta.8) (2020-07-07)


### Bug Fixes

* **object:** extend from Record ([6c7e60e](https://github.com/schuchard/taap/commit/6c7e60e00587966e2ff3424e20f7495e094eee3d))

# [1.0.0-beta.7](https://github.com/schuchard/taap/compare/v1.0.0-beta.6...v1.0.0-beta.7) (2020-07-06)


### Features

* **ci:** dist files ([ed3e039](https://github.com/schuchard/taap/commit/ed3e039b4de022b7c9f43991faa536c5d04260b3))

# [1.0.0-beta.6](https://github.com/schuchard/taap/compare/v1.0.0-beta.5...v1.0.0-beta.6) (2020-07-06)


### Features

* fix build command ([8eebde6](https://github.com/schuchard/taap/commit/8eebde6b071130355e9000140cf6f9f3f9adf4d0))

# [1.0.0-beta.5](https://github.com/schuchard/taap/compare/v1.0.0-beta.4...v1.0.0-beta.5) (2020-07-06)


### Features

* build ([31a5ddb](https://github.com/schuchard/taap/commit/31a5ddbc0956819e4d209fc3f3ffa62cfd1ef360))
* **ci:** npm changes ([9d546cf](https://github.com/schuchard/taap/commit/9d546cf9d28a2ad6917dda0ba4424af2ca9c542f))

# [1.0.0-beta.4](https://github.com/schuchard/taap/compare/v1.0.0-beta.3...v1.0.0-beta.4) (2020-07-06)


### Features

* **ci:** include build step in release ([0dd3952](https://github.com/schuchard/taap/commit/0dd3952382fc062bf4f1eb284d2e6d7c67d8b607))

# [1.0.0-beta.3](https://github.com/schuchard/taap/compare/v1.0.0-beta.2...v1.0.0-beta.3) (2020-07-06)


### Features

* **ci:** add new glob pattern for dist folder ([b22dc8a](https://github.com/schuchard/taap/commit/b22dc8a16dbd11b184c4b8879d1297d9c8b28627))

# [1.0.0-beta.2](https://github.com/schuchard/taap/compare/v1.0.0-beta.1...v1.0.0-beta.2) (2020-07-06)


### Features

* **ci:** add dist folder ([39b90e4](https://github.com/schuchard/taap/commit/39b90e443407d4ae28807c031a232d71c203c27e))

# 1.0.0-beta.1 (2020-07-06)


### Features

* init and update project, add src files and specs ([6d6cc15](https://github.com/schuchard/taap/commit/6d6cc15767bfbfef82c2bda1de0d78843fc76650))


test/taap.test.ts
import {
  isArray,
  isObject,
  isString,
  isDate,
  isRegExp,
  isFunction,
  isBoolean,
  isNumber,
  isBigInt,
  isNull,
  isUndefined,
  getType,
} from '../src/taap';

const regExp = /\w+\s/g;

describe('isArray', () => {
  it('should return a true if an Array', () => {
    expect(isArray([])).toBe(true);
  });

  it('should assert the type predicate and compile', () => {
    const maybeArray: unknown = [1];

    if (isArray<number>(maybeArray)) {
      maybeArray.push(1);
      expect(maybeArray[1]).toEqual(1);
    }

    expect.assertions(1);
  });
});

describe('isObject', () => {
  it('should return a true if an Object', () => {
    expect(isObject({})).toBe(true);
  });

  it('should assert the type predicate and compile', () => {
    const maybeObject: unknown = { a: 1 };

    if (isObject<Record<string, number>>(maybeObject)) {
      maybeObject.b = 2;
      expect(Object.keys(maybeObject).map((v) => v)).toEqual(['a', 'b']);
      expect(Object.values(maybeObject).map((v) => v)).toEqual([1, 2]);
    }
    expect.assertions(2);
  });
});

describe('isString', () => {
  it('should return a true if a String', () => {
    expect(isString('')).toBe(true);
  });

  it('should assert the type predicate and compile', () => {
    let maybeString = 'a,b' as unknown;

    if (isString(maybeString)) {
      maybeString += ',c';
      expect(maybeString.split(',')).toEqual(['a', 'b', 'c']);
    }
    expect.assertions(1);
  });
});

describe('isDate', () => {
  it('should return a true if a Date', () => {
    expect(isDate(new Date())).toBe(true);
  });

  it('should assert the type predicate and compile', () => {
    const maybeDate: unknown = new Date('August 19, 1975 23:15:30');

    if (isDate(maybeDate)) {
      maybeDate.setFullYear(1969);
      expect(maybeDate.toDateString()).toEqual('Tue Aug 19 1969');
    }

    expect.assertions(1);
  });
});

describe('isRegExp', () => {
  it('should return a true if a RegExp', () => {
    expect(isRegExp(/test/i)).toBe(true);
  });

  it('should assert the type predicate and compile', () => {
    const maybeRegExp: unknown = regExp;

    if (isRegExp(maybeRegExp)) {
      const str = 'fee fi fo fum';
      expect(str.match(maybeRegExp)).toEqual(['fee ', 'fi ', 'fo ']);
      expect(maybeRegExp.test('fee fi fo fum')).toEqual(true);
    }

    expect.assertions(2);
  });
});

describe('isFunction', () => {
  it('should return a true if a Function', () => {
    expect(isFunction(() => ({}))).toBe(true);
  });

  it('should assert the type predicate and compile', () => {
    const maybeFunction: unknown = (x: number) => x.toString();

    if (isFunction<(x: number) => string>(maybeFunction)) {
      expect(maybeFunction(42)).toEqual('42');
    }

    expect.assertions(1);
  });
});

describe('isBoolean', () => {
  it('should return a true if a Boolean', () => {
    expect(isBoolean(true)).toBe(true);
  });

  it('should assert the type predicate and compile', () => {
    const maybeBoolean: unknown = true;

    if (isBoolean(maybeBoolean)) {
      expect(maybeBoolean).toBe(true);
    }

    expect.assertions(1);
  });
});

describe('isNumber', () => {
  it('should return a true if a Number', () => {
    expect(isNumber(1)).toBe(true);
  });

  it('should assert the type predicate and compile', () => {
    const maybeNumber: unknown = 1;

    if (isNumber(maybeNumber)) {
      expect(maybeNumber + 2).toEqual(3);
    }

    expect.assertions(1);
  });
});

describe('isBigInt', () => {
  it('should return a true if a BigInt', () => {
    expect(isBigInt(BigInt(42))).toBe(true);
  });

  it('should assert the type predicate and compile', () => {
    const maybeBigInt: unknown = BigInt(42);

    if (isBigInt(maybeBigInt)) {
      // todo target tests to ES2020 to use BigInt's: `42n`
      expect(maybeBigInt.toString()).toEqual('42');
    }

    expect.assertions(1);
  });
});

describe('isNull', () => {
  it('should return a true if an Null', () => {
    expect(isNull(null)).toBe(true);
  });

  it('should assert the type predicate and compile', () => {
    const maybeNull: unknown = null;

    if (isNull(maybeNull)) {
      expect(maybeNull).toEqual(null);
    }

    expect.assertions(1);
  });
});

describe('isUndefined', () => {
  it('should return a true if an Undefined', () => {
    expect(isUndefined(undefined)).toBe(true);
  });

  it('should assert the type predicate and compile', () => {
    const maybeUndefined: unknown = undefined;

    if (isUndefined(maybeUndefined)) {
      expect(maybeUndefined).toEqual(undefined);
    }

    expect.assertions(1);
  });
});

describe('getType', () => {
  test.each`
    kind           | value
    ${'Array'}     | ${[]}
    ${'Array'}     | ${[1]}
    ${'Array'}     | ${[{ a: 1 }]}
    ${'Array'}     | ${[false]}
    ${'Object'}    | ${{}}
    ${'Object'}    | ${{ a: 1 }}
    ${'String'}    | ${''}
    ${'String'}    | ${'taap'}
    ${'Date'}      | ${new Date()}
    ${'RegExp'}    | ${regExp}
    ${'RegExp'}    | ${new RegExp('a')}
    ${'Boolean'}   | ${true}
    ${'Boolean'}   | ${false}
    ${'Boolean'}   | ${!!1}
    ${'Number'}    | ${1}
    ${'Number'}    | ${2_000}
    ${'Number'}    | ${0.4}
    ${'Number'}    | ${-100}
    ${'Null'}      | ${null}
    ${'Undefined'} | ${undefined}
    ${'BigInt'}    | ${BigInt(9007199254740991)}
    ${'Function'}  | ${() => ({})}
    ${'Function'} | ${function testFunc() {
  return false;
}}
  `('returns $kind when given $value', ({ kind, value }) => expect(getType(value)).toEqual(kind));
});


rollup.config.ts
import resolve from 'rollup-plugin-node-resolve'
import commonjs from 'rollup-plugin-commonjs'
import sourceMaps from 'rollup-plugin-sourcemaps'
import camelCase from 'lodash.camelcase'
import typescript from 'rollup-plugin-typescript2'
import json from 'rollup-plugin-json'

const pkg = require('./package.json')

const libraryName = 'taap'

export default {
  input: `src/${libraryName}.ts`,
  output: [
    { file: pkg.main, name: camelCase(libraryName), format: 'umd', sourcemap: true },
    { file: pkg.module, format: 'es', sourcemap: true },
  ],
  // Indicate here external modules you don't wanna include in your bundle (i.e.: 'lodash')
  external: [],
  watch: {
    include: 'src/**',
  },
  plugins: [
    // Allow json resolution
    json(),
    // Compile TypeScript files
    typescript({ useTsconfigDeclarationDir: true }),
    // Allow bundling cjs modules (unlike webpack, rollup doesn't understand cjs)
    commonjs(),
    // Allow node_modules resolution, so you can use 'external' to control
    // which external modules to include in the bundle
    // https://github.com/rollup/rollup-plugin-node-resolve#usage
    resolve(),

    // Resolve source maps to the original source
    sourceMaps(),
  ],
}


tools/gh-pages-publish.ts
const { cd, exec, echo, touch } = require("shelljs")
const { readFileSync } = require("fs")
const url = require("url")

let repoUrl
let pkg = JSON.parse(readFileSync("package.json") as any)
if (typeof pkg.repository === "object") {
  if (!pkg.repository.hasOwnProperty("url")) {
    throw new Error("URL does not exist in repository section")
  }
  repoUrl = pkg.repository.url
} else {
  repoUrl = pkg.repository
}

let parsedUrl = url.parse(repoUrl)
let repository = (parsedUrl.host || "") + (parsedUrl.path || "")
let ghToken = process.env.GH_TOKEN

echo("Deploying docs!!!")
cd("docs")
touch(".nojekyll")
exec("git init")
exec("git add .")
exec('git config user.name "kevin schuchard"')
exec('git config user.email "schuchard.kevin@gmail.com"')
exec('git commit -m "docs(docs): update gh-pages"')
exec(
  `git push --force --quiet "https://${ghToken}@${repository}" master:gh-pages`
)
echo("Docs deployed!!")


src/taap.ts
export type TaapType =
  | 'Array'
  | 'Object'
  | 'String'
  | 'Date'
  | 'RegExp'
  | 'Function'
  | 'Boolean'
  | 'Number'
  | 'BigInt'
  | 'Null'
  | 'Undefined';

export function getType(v: any) {
  return Object.prototype.toString.call(v).slice(8, -1) as TaapType;
}
/**
 * Determine if argument is an `Array`.
 *
 * ```ts
 * const maybeArray: unknown = [1];
 *
 * if (isArray(maybeArray)) {
 *   // `maybeArray` will be considered an Array
 *   maybeArray.map(x => x)
 * }
 * ```
 * Provide a specific type:
 * ```ts
 * const maybeArray: unknown = [1];
 *
 * if (isArray<number>(maybeArray)) {
 *   // `maybeArray` will be considered an Array and `x` will be considered a number
 *   maybeArray.map((x => x) // `
 * }
 * ```
 *
 * @typeParam T Accepts a type to narrow the return type.
 *
 * Default: to `any`.
 */
export function isArray<T = any>(val: any): val is Array<T> {
  return getType(val) === 'Array';
}

/**
 * Determine if argument is an `Object`.
 *
 * ```ts
 * const maybeObject: unknown = { a: true };
 *
 * if (isObject(maybeObject)) {
 *   // `maybeObject` will be considered a Record<string, any>
 *   maybeObject.someProp = true
 * }
 * ```
 * Provide a specific type:
 * ```ts
 * const maybeObject: unknown = { a: true };
 *
 * if (isObject<{a: boolean}>(maybeObject)) {
 *   // maybeObject will be typed as `{ a: boolean }`
 *   maybeObject.a === true;
 * }
 * ```
 *
 * @typeParam T Accepts a type to narrow the return type.
 */
export function isObject<T extends Record<string, any>>(val: any): val is T {
  return getType(val) === 'Object';
}

/**
 * Determine if argument is a `Function`.
 *
 * ```ts
 * const maybeFunction: unknown = (x: any) => x;
 *
 * if (isFunction(maybeFunction)) {
 *   maybeFunction.call(...args)
 * }
 * ```
 *
 * Provide a specific type:
 * ```ts
 * const maybeFunction: unknown = (x: number) => x.toString();
 *
 * if (isFunction<(x: number) => string>(maybeFunction)) {
 *   maybeFunction(42) === '42'
 * }
 * ```
 *
 * @typeParam T Accepts a type to narrow the return type.
 */
export function isFunction<T = () => {}>(val: any): val is T {
  return getType(val) === 'Function';
}

/**
 * Determine if argument is a `string`.
 *
 * ```ts
 * const maybeString: unknown = 'ta,ap';
 *
 * if (isString(maybeString)) {
 *   maybeString.split(',').join('') === 'taap'
 * }
 * ```
 */
export function isString(val: any): val is string {
  return getType(val) === 'String';
}

/**
 * Determine if argument is a `Date`.
 *
 * ```ts
 * const maybeDate: unknown = new Date('August 19, 1975 23:15:30');
 *
 * if (isDate(maybeDate)) {
 *   maybeDate.setFullYear(1969);
 * }
 * ```
 */
export function isDate(val: any): val is Date {
  return getType(val) === 'Date';
}

/**
 * Determine if argument is a `RegEx`.
 *
 * ```ts
 * const maybeRegExp: unknown = /\w+\s/g;
 *
 * if (isRegExp(maybeRegExp)) {
 *   const str = 'fee fi fo fum';
 *   maybeRegExp.test('fee fi fo fum') === true
 * }
 * ```
 */
export function isRegExp(val: any): val is RegExp {
  return getType(val) === 'RegExp';
}

/**
 * Determine if argument is a `boolean`.
 *
 * ```ts
 * const maybeBoolean: unknown = true;
 *
 * if (isBoolean(maybeBoolean)) {
 *   maybeBoolean === true
 * }
 * ```
 */
export function isBoolean(val: any): val is boolean {
  return getType(val) === 'Boolean';
}

/**
 * Determine if argument is a `number`.
 *
 * ```ts
 * const maybeNumber: unknown = 1;
 *
 * if (isNumber(maybeNumber)) {
 *   maybeNumber + 2 === 3;
 * }
 * ```
 */
export function isNumber(val: any): val is number {
  return getType(val) === 'Number';
}

/**
 * Determine if argument is a `BigInt`.
 *
 * ```ts
 * const maybeBigInt: unknown = BigInt(42);
 *
 * if (isBigInt(maybeBigInt)) {
 *   maybeBigInt === 42n;
 * }
 * ```
 */
export function isBigInt(val: any): val is BigInt {
  return getType(val) === 'BigInt';
}

/**
 * Determine if argument is `null`.
 *
 * ```ts
 * const maybeNull: unknown = 1;
 *
 * if (isNumber(maybeNull)) {
 *   maybeNull === null;
 * }
 */
export function isNull(val: any): val is null {
  return getType(val) === 'Null';
}

/**
 * Determine if argument is `undefined`.
 *
 * ```ts
 * const maybeUndefined: unknown = 1;
 *
 * if (isNumber(maybeUndefined)) {
 *   maybeUndefined === undefined;
 * }
 */
export function isUndefined(val: any): val is undefined {
  return getType(val) === 'Undefined';
}
